### ISSUE TITLE
repeat chart with top level defined brush selection corrupts cross-view interaction

### ISSUE THREAD

**Issue by mattijn** [vega/vega-lite#8348]
This issue is linked to this comment: https://github.com/vega/vega-lite/issues/8230#issuecomment-1201358652.
Example is based on a repeated chart such as this example: https://vega.github.io/vega-lite/examples/interactive_layered_crossfilter.html

Upon manually defining a `"name":"CHART"` property for the view defining the `param` + `filter` the units-field in the `brush_store` is fixed to the name `"CHART"`, as can be seen in the data-viewer below. No matter which view is being brushed. This makes the brush selection corrupt cross-view.

<img width="700" alt="image" src="https://user-images.githubusercontent.com/5186265/183510923-eefe892e-a9ec-4e04-916e-d644d004efc3.png">

Spec that errors:
[Open the Chart in the Vega Editor](https://vega.github.io/editor/#/url/vega-lite/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykBaADZ04JAKyUAVhDYA7EABoQAEySYUqUAwBOgtCrVICAM2FksEfgCYA1lJnylxttuSY0oHEm0Q4Hg5h+6KqBIAC+EUqY2kiyEM6uaADaoFBIglAMgmpBGGw6EAAUIQwIlCFwAJSKIEgQ+pg0CH5hALpK2nA4cGr+UGyCpfKoSSo0EOqyUH5KynDZAJ41jc0grWFKEN1Q-gDuNMr0aACMYgAMSvA05u6ox2cXIItw2smgyNo2+kzeNXBTbGUNFkZH8AA9-MYaPNlP5Ot1euh+oMEPINiAmMD-MgwZi4mgrGdIiAlppamQyJ0KKEkflZO50ciXP5iOkGLkAMQANh54Uinm8SAQ9RGoFkQtyTG0DAgMBqvkEcCgt1AmAW3X0wMC2lZeiU-36QJBItGEPWbSiMTiCQQbxAUME2v8XhitvQUplcoi7RA4tW6AAwgAJACCACUACo1D5fd2-fUAo2gskQslQmFwro9W4gZFDcJKPHYpC44EiwnE0mgJAUqk5fT9Bj0gu5gbMsm69n6FyxMgtb0RIA)

* * *

Comparison to _not_ using a manual defined `name` property:

When I make a brush selection the first view (most left), I observe that the `unit` field is filled with `"child__column_distance_layer_0"` and when changing the brush selection to the middle-plot the `unit` field changes to `"child__column_delay_layer_0"` and the last view has an unit in the brush store defined as `"child__column_time_layer_0"`.

<img width="700" alt="image" src="https://user-images.githubusercontent.com/5186265/183509532-e95ad236-460c-4f4b-9553-a26ee3d04b2f.png">

I somehow hoped I could use these programmatically defined units as different `views` in a top-level defined `params` object as such:

```json
"params": [
  {
    "name": "brush",
    "select": {"type": "interval", "encodings": ["x"]},
    "views": ["child__column_distance_layer_0", "child__column_delay_layer_0", "child__column_time_layer_0"]
  }
]
```
but to no avail (`[Error] Unrecognized signal name: "brush"`)..

**Comment by mattijn** [comment]
In this comment by @arvind  https://github.com/vega/vega-lite/issues/8230#issuecomment-1179213186 is mentioned: 
> [a `view`] can either be a complete name of a view or a partial path through the view tree. 

I was wondering if this _partial path through the view tree_ can be of usage here? 

If there is some documentation available on these _partial path_ or if someone can give some insights than I can investigate by myself a bit more as well.

**Comment by mattijn** [comment]
Following the described logic from here, https://github.com/vega/vega-lite/issues/8452#issuecomment-1279995980, I think I followed the _partial path through the view tree_ correctly, but I was not able to succeed. Reading through this issue, I've tried it before, but without a spec that can be directly opened in the editor, now one can: [Open the Chart in the Vega Editor](https://vega.github.io/editor/#/url/vega-lite/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykBaADZ04JAKyUAVhDYA7EABoQAEySYUqUAwBOgtCrVICAM2FksEfgCYA1lJnylxttuSY0oHEm0Q4Hg5h+6KqBIAC+EUpe2kgIEGgA2qCysUEgTNoMEDCKIL6CcFDumiCYAJ44aTSygdrESHpKcLJQbMrVZPGoCSAAHiAAumFKxDRwAO5dPbA0gsoA+vOtggwIsvPtEOotcPOCSGVw2vMADLkzc4vLq+vKcPtlewdHp+cwswtLbCtr85g0CF2DxeZyGAyUmBisggzlciVAUAaUAY+1C6BgbB0EAAFCFVpQQnAAJS5JBdUoAvxhcEgbRwSpqfzXNaJFQ0LZIHa5O4PXL-QGDYZ5SpQfzjGjKehoACMYhOSngNHMxWlJ3lIGB2nhIGQ2hs+iY3lyzVa7VkZH8-RKxjGc38dIZxRAzPkQqY1X8yF67uhaCsJ0iIDK-iQZDIdIoaOdmJq4SUyxc-nqKzSAGIAGyZ8KRUCQzkwlwIbU2wS1fzRWIGzLZbM03X69CGrVNFptDqW-w2+7Ke30uCM9AuuPpD0lL0+rr+wPBkqh8N8KOtBixoUJrUlZMMNIuTlkKkRIZhIA).
The error still is: 

> ```cmd
> Unrecognized signal name: "brush"
> ```

**Comment by arvind** [comment]
Hi @mattijn, 

There are two things at play here that are preventing the spec from working correctly:

(1) Neither of the layers have a `name` property, which does not allow them to be addressed as `views` in top-level selection parameters. 

(2) To address a nested view, `views` needs to be an array of names rather than concatenating the names together. 

[Here's a working spec](https://vega.github.io/editor/#/url/vega-lite/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykBaADZ04JAKyUAVhDYA7EABoQAEySYUqUAwBOgtCrVICAM2FksEfgCYA1lJnylxttuSY0oHEm0Q4Hg5h+6KqBIAC+EUpe2kgIEGgA2qCysUEgTNoMEDCKIL6CcFDumiCYAJ44aTSygdrESHpKcLJQbMrVZPGoCSAAHiAAumFKxDRwAO5dCT2wNILKAPoLrYIMCLIL7RDqLX5KgkhlcNoLAAyDCjMwc4vLbKvrm3AHZbkvx2cXVzdLK2sbmBoCD2IHeJ3OAyGAyUmBisggzlciVAUAaUAYB1C6BgbB0EAAFCE1pQQnAAJS5JBdUpAvxhaEgbRwSpqfx-daJFQ0bZIXa5ZTPQ65QHAwbDPKVKD+cY0ZT0NAARjEpyU8Bo5mKCtOKtBh2OyJAKVF6DBnyUyG0Nn0TG8uWarXasjI-n6JWMY3m-iZLOKIHZ8nFTGq-mQvSD8LQVlOkRArxKSDIZCZFCxftxNXCShWLn89VWaQAxAA2EvhSKgWG8hEuBAG92CWr+aKxa2ZbJlhkWq3oG3aO0tNodF3+d3PZRe5lwVnof2Z9LBkqh8NdKMxuOgBNJvip1oMDPi7N9kp5hhpFy8sh0iJDMJAA)

(There technically is a bug that causes duplicate names to be added if the top-level view is named (as is the case here since the top-level view is a repeated view). This bug should be fixed once #8486 is merged. But, as far as I can tell, this bug doesn't prevent this spec from working.)

Hope this helps.

**Comment by arvind** [comment]
Actually, I'm reopening this because the "working spec" I linked to above is showing the behavior you screenshotted in the OP. It seems like multiple brushes appear even though there should only be one.

**Comment by mattijn** [comment]
Thanks, as always @arvind! Your 2nd listed point is eye-opening:

> (2) To address a nested view, `views` needs to be an array of names rather than concatenating the names together.

I tried lots of things, but this not. Sometimes I feel like being a VL-philologist;).

Regardless the outstanding behavior, which is almost as it should be (stay positive), it's good to know that we can create the right specification with top-declared parameters using this logic.

**Comment by ChristopherDavisUCI** [comment]
Hi @mattijn, I made a first attempt at fixing this problem here: https://github.com/vega/vega-lite/commit/0ab291a413c58728212c2a9570afbb6b99185454

The resulting extended Vega-Lite for your first chart above is this: [Open the Chart in the Vega Editor](https://vega.github.io/editor/#/url/vega-lite/N4IgJghgLhIFygK4CcA29zQgegGaoEsBzACygGcBaAJgGsA6AK3IHsA7EAGhFxeQFto8UAAcIycgFNhmKNLizpAXxXcIhIhwXr03ACTkAxiUmCMZKCPJxs2AG6SiEekQJQSiAEb0CLbEZNBe0cISkI5ewBWJlYObihkCDZyXgF4AG1QQ3VDRFRoeRASFhRyAApIKER+ekrJAEouEAhyDCgCfmUAXW5DFlRq5PgAZl72bKgM0AB3AjB3eABGSIAGbhNiMiWVtZB8gE9JZCmQQWRaDE9xJsk2PrACNiIZAA8ZXAJJVDAMB-IYO7SbieR4yQQvEFDODUFaqED7GQQIhEZAhOQYPqINiTJRjVB8GR2dSIQoAYgAbJSQCpcaJxBB+K04JkQGwGYVPMhEOQSE0pKhJIZJggQFB9iJCo85MgiboQLd7o8iEz0iA3l0VD1WeyMSQCN8APoGvoDfhsA1-AGGSQAYQAEgBBABKABUmmcLgorsduAqWA8nq93p9vr8CP8ktampCwRAIY8mTC4QiRUiUWjCpjsdS8QSRbKSRg+EkiMpcaLEslUvwTh9UNKZGJEjWvVyedSlBqtWzOrr9WAjSbBhbw1blJwZnMFnBlrsNqRhYsdtwDkcTh7LtdfXd-UqgyKPl8fgowF8IAjgaCReDIYnYeWU6A06inOiFFmcbnjvniWTKeSO1pEAmwZFVQB7Dk215bh+UFYVQDFCUMClI5ZRuHcA2VDI1RADVO24CC+0NY1+mHU8DntZ03W4DcvS3eUML3EU3gPENj3AM8LxAGNrzjW94CTB9EWRF8CgxEps3LE081AAtCmLJ4y3iSsUj4FsWTrBsRRAltuKgjsuwInV3z1YihzNC1OJzSd5l5GdVnWSRNkXZc9nPNdmVAWjuPov1MP3UBD1DBR2l7S8tC8viEwE+9uEfZoRIzcSsU-EBpO-WTfwwCkqRpCdgPpRkTkI1tuWgkBYKFGREMlbFUPUdDFSeFUcLw7tjLS0yBxI01zVC21HVdd1xE9HyfQYprnmY4MjzaDogW4q9IvjKFBLi4T01fTMJNS9LCSyhQFNLDtlKSVS0k8nh9S0ulm0ufTNXw7VexM-tB1Iiz+upDUgA). Does that seem to be working as expected?

**Comment by mattijn** [comment]
Yes, your example works correctly, but that is because the `params` are localized and not defined top-level like we do in Altair isn't it?

How is this example related to the change you introduced that hopefully fix the issue (would be great btw!)?

Probably not related to this issue: I also noticed a top-level `"align": "all"` key-val in the spec. I can't recall seeing that before. Did you deliberately add this?

**Comment by ChristopherDavisUCI** [comment]
Hi @mattijn, I think these changes you mention always(?) happen in the transition to the "extended Vega-Lite spec" (visible at the bottom of the Vega editor).  For example, I think they also happen in your example at the top.  The only change I made was to add a prefix to certain names, replacing `spec.name` by `params.repeaterPrefix + spec.name` when both `spec.name` and `params.repeaterPrefix` are defined.

What I linked in https://github.com/vega/vega-lite/issues/8348#issuecomment-1399378752 was the extended Vega-Lite spec, but the original Vega-Lite spec that I entered was the same as yours.

I might make a PR with this change just to see what the feedback is.  I don't think this exact change to the name is what they will want, but if we're lucky maybe the location of the change and/or the logic is correct.

**Comment by mattijn** [comment]
I was not aware that the extended Vega-Lite spec is doing some translation of the top-level parameters to nested parameters. I don't think the used spec in OP is valid to the current issue we are seeing.

Does your fix also works for this spec: [Open the Chart in the Vega Editor](https://vega.github.io/editor/#/url/vega-lite/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykBaADZ04JAKyUAVhDYA7EABoQAEySYUqUAwBOgtCrVICAM2FksEfgCYA1lJnylxttuSY0oHEm0Q4Hg5h+6KqBIAC+EUpe2kgIEGgA2qCysUEgTNoMEDCKIL6CcFDumiCYAJ44aTSygdrESHpKcLJQbMrVZPGoCSAAHiAAumFKxDRwAO5dCT2wNILKAPoLrYIMCLIL7RDqLX5KgkhlcNoLAAyDCjMwc4vLbKvrm3AHZbkvx2cXVzdLK2sbmBoCD2IHeJ3OAyGAyUmBisggzlciVAUAaUAYB1C6BgbB0EAAFCE1pQQnAAJS5JBdUpAvxhaEgbRwSpqfx-daJFQ0bZIXa5ZTPQ65QHAwbDPKVKD+cY0ZT0NAARjEpyU8Bo5mKCtOKtBh2OyJAKVF6DBnyUyG0Nn0TG8uWarXasjI-n6JWMY3m-iZLOKIHZ8nFTGq-mQvSD8LQVlOkRArxKSDIZCZFCxftxNXCShWLn89VWaQAxAA2EvhSKgWG8hEuBAG92CWr+aKxa2ZbJlhkWq3oG3aO0tNodF3+d3PZRe5lwVnof2Z9LBkqh8NdKMxuOgBNJvip1oMDPi7N9kp5hhpFy8sh0iJDMJAA) from this [comment](https://github.com/vega/vega-lite/issues/8348#issuecomment-1280162030) of @arvind? Because that spec is actually defining top-level params as what we are intend to do in Altair as well.

Another thing I notice from @arvind spec is that it is using these concatenation of names as arrays with defined layer:
```json
"views": [
  ["child__column_distance", "layer_0"],
  ["child__column_delay", "layer_0"],
  ["child__column_time", "layer_0"]
]
```
But If I look to what type of Vega-lite we generate from the following Altair spec than I don't see this is happening:
```python
source = alt.UrlData(
    data.flights_2k.url,
    format={'parse': {'date': 'date'}}
)

brush = alt.selection(type='interval', encodings=['x'])

# Define the base chart, with the common parts of the
# background and highlights
base = alt.Chart(source).mark_bar().encode(
    x=alt.X(alt.repeat('column'), type='quantitative', bin=alt.Bin(maxbins=20)),
    y='count()'
).properties(
    width=160,
    height=130
)

# gray background with selection
background = base.encode(
    color=alt.value('#ddd')
).add_params(brush)

# blue highlights on the transformed data
highlight = base.transform_filter(brush)

# layer the two charts & repeat
chart = alt.layer(
    background,
    highlight,
    data=source
).transform_calculate(
    "time",
    "hours(datum.date)"
).repeat(column=["distance", "delay", "time"])
```
In my case it is just:
```json
"views": ["view_9"]
```
Did we ever implemented this? Developments are going so quick last few weeks, I can't remember it anymore.

**Comment by ChristopherDavisUCI** [comment]
Thanks @mattijn!  No, that spec you linked is not working for me either, but it gives a different error, `Unrecognized signal name: "brush"`, so maybe that is a sign of progress.  (The error shows up for me right away, as opposed to the current version of Vega-Lite, where the error doesn't show up until you attempt to use the parameter.)

My memory is also a little fuzzy about the special names, but I remember we did implement at least some of it here:
https://github.com/altair-viz/altair/pull/2684/commits/f5473234a43b054e778541665544086895d65d9f
So far I haven't succeeded in finding an example where this code actually gets used, but I will keep looking!

**Comment by mattijn** [comment]
Isn't this the example where that code should be triggered?

**Comment by ChristopherDavisUCI** [comment]
Maybe?!  But my inclination is always to give the benefit of the doubt to my past self, not to my current self who has looked at it for 2 minutes 😝

**Comment by mattijn** [comment]
Yes, my current self is also much more confused than the moment we implemented that part😀

**Comment by ChristopherDavisUCI** [comment]
I'll try to look over this stuff carefully in the next few days.

**Comment by mattijn** [comment]
@ChristopherDavisUCI I opened https://github.com/altair-viz/altair/issues/2849 to track the Altair issue described above. One more thing, you mentions the following:

> _The error shows up for me right away, as opposed to the current version of Vega-Lite, where the error doesn't show up until you attempt to use the parameter._

But the current version does not give an error, but a warning:
```cmd
[Warning] Infinite extent for field "__count": [Infinity, -Infinity]
```
So the interaction is still doing something, but just something inaccurate.

**Comment by ChristopherDavisUCI** [comment]
@mattijn I took a closer look at Arvind's example you linked in https://github.com/vega/vega-lite/issues/8348#issuecomment-1399528866.  In that example, we have
```
"views": [
        ["child__column_distance", "layer_0"],
        ["child__column_delay", "layer_0"],
        ["child__column_time", "layer_0"]
      ]
```
In the resulting `concat` chart, three views with the name `"layer_0"` appear.  My strategy in #8663 is to give those views different names.  When I try Arvind's spec (with PR #8663) with the following change it works:
```
"views": [
        "child__column_distancelayer_0",
        "child__column_delaylayer_0",
        "child__column_timelayer_0"
      ]
```
Any first impressions on that?

It doesn't seem worthwhile making a corresponding fix on the Altair side, because even if this is a good strategy, I doubt the exact naming I did (just concatenating the name and the prefix) is what should be done on the Vega-Lite side.

**Comment by mattijn** [comment]
I just noticed your comment @ChristopherDavisUCI, but
```
["child__column_distance", "layer_0"],
```
Isn't equal to
```
"child__column_distancelayer_0",
```
Am I right in understanding your comment is that the currently defined altair-views are resolving, or at least should resolve into the above in the _extended vega lite specification_ on the VL-side?

**Comment by ChristopherDavisUCI** [comment]
Thanks for checking it out @mattijn.  Think of my comment above as having nothing to do with Altair at this point.  (My idea is that once the issue from Arvind's chart is resolved on the Vega-Lite side, we can hopefully work something out to match it on the Altair side.)  In #8663 I give unique names to the various views and that seemed to get the chart working.  (I didn't put much thought into what exact names to use, figuring I didn't know what conventions, like how many underscores, should be used.  All I did was prepend the repeaterPrefix onto the beginning of the current name.)

Does that make sense?  I believe as things currently are defined in Vega-Lite, in the extended specs, some views can have the same name (and are only distinguished by having a different path through the view tree).

**Comment by mattijn** [comment]
Ok! More clear now. I'll subscribe to #8663 to see how this lands in VL👍

**Comment by mattijn** [comment]
If I'm not misunderstanding things, than this issue is fixed by #8733, see https://github.com/vega/vega-lite/pull/8733#issuecomment-1451337875.
Please re-open if I'm wrong on this.
